### PR TITLE
add alacritty

### PR DESCRIPTION
**Pull request by exi**
###### Motivation for this change
Add alacritty.

This package will only build once https://github.com/NixOS/nixpkgs/pull/21800 is merged

### CODE REVIEW

**Comment by exi** [comment]
@globin some love?

File: pkgs/applications/misc/alacritty/default.nix [hunk 1]
```diff
@@ -0,0 +1,42 @@
+{ stdenv, fetchFromGitHub, rustPlatform, cmake, gcc, pkgconfig,
+  freetype, expat, gperf, libX11, libXcursor, libXxf86vm, libXi }:
+
+with rustPlatform;
+
+buildRustPackage rec {
+  name = "alacritty-${version}";
```
Comment (Mic92) [inline pkgs/applications/misc/alacritty/default.nix]: usually we have the following scheme when packaging packages without releases:

```
  name = "alacritty-unstable-<year>-<month>-<day>";
```

File: pkgs/applications/misc/alacritty/default.nix [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    repo = "alacritty";
+    rev = "f4b10a1dbb991861a84c5085a0ee4384d2377448";
+    sha256 = "044l5s1w2g5krngr1v4lfxv7smh6wzy9zrabnihnpf1bsg6q73nb";
+  };
+
+  nativeBuildInputs = [
+    cmake gcc pkgconfig freetype expat gperf
+    libX11 libXcursor libXxf86vm libXi
+  ];
+  buildInputs = [ ];
```
Comment (Mic92) [inline pkgs/applications/misc/alacritty/default.nix]: that could be dropped then.

**Comment by gebner** [comment]
Should we make hydra build this package?  Right now we don't build rustNightly (or rustBeta), see #18111.  cc @the-kenny

File: pkgs/applications/misc/alacritty/default.nix (hunk 1 above)
Comment (exi) [inline pkgs/applications/misc/alacritty/default.nix]: fixed

File: pkgs/applications/misc/alacritty/default.nix (hunk 2 above)
Comment (exi) [inline pkgs/applications/misc/alacritty/default.nix]: fixed

**Review by puffnfresh — APPROVED** [review]

**Review by jb55 — APPROVED** [review]

**Comment by puffnfresh** [comment]
I was unable to build locally due to the Rust compiler stopping compilation with:

> udp_clone_smoke has been running for over 60 seconds

But was able to build by overriding it with `rustPlatform = (makeRustPlatform rustNightlyBin);` - tried it out, it runs well!

**Comment by exi** [comment]
@puffnfresh i suspect that this happens because this branch is not rebased onto the latest master.
I just rebased and force pushed it and will try the build again myself.

**Comment by exi** [comment]
@puffnfresh just tried it with this branch on top of the current master and rust beta, rust nightly and alacritty build without issues

**Comment by jb55** [comment]
I can't seem to build either, I'm getting this: https://gist.github.com/jb55/7c438c1ecbb6ee34551aeb771e4841fe

**Comment by exi** [comment]
@jb55 you are not building from the right branch, the nightly version this is trying to build is way too old.
did you forget to "git pull" your master or something?

**Comment by jb55** [comment]
oops looks like my nixpkgs was dirty and had stuff in staging area. retrying now

**Comment by jb55** [comment]
works great :+1:

**Comment by puffnfresh** [comment]
I'm getting the same UDP test problems so it must be something to do with my system.

**Comment by Mic92** [comment]
@puffnfresh what operating system do you use? I had problems with failing udp tests on debian in a nix-user-chroot, but not on nixos.

**Comment by puffnfresh** [comment]
@Mic92 I use NixOS

**Comment by exi** [comment]
@puffnfresh @Mic92 could you check out the latest master and try again? I am wondering if this was an issue with the parallel/codegen PR that has been reverted 3 days ago

**Comment by puffnfresh** [comment]
I merged master into this branch and tried to build but I got the hang again:

> test net::udp::tests::udp_clone_smoke has been running for over 60 seconds

**Comment by exi** [comment]
@puffnfresh this seems like a bug that might be worth raising as a separate issue.

**Comment by Mic92** [comment]
sandbox build failed because a dependency was attempt to download from git.

```
warning: custom registry support via the `registry.index` configuration is being removed, thisfunctionality will not work in the future
    Updating git repository `https://github.com/jwilm/glutin`
error: failed to load source for a dependency on `glutin`
```

**Comment by the-kenny** [comment]
@Mic92 this happens because `alacritty` uses a git-dependency of `glutin`. Not supported (yet?) on our build infrastructure. I think this error is caused by `cargo` not downloading git repositories in a non-deterministic way.

One would need to pin that revision and provide a separate `fetchgit` to download it. Alternative would be fixing cargo to use `git archive` or something instead of just cloning the repo.

**Comment by Mic92** [comment]
@the-kenny actually `buildRustPackage` support that, but it seems that cargo tries to update the pre-fetched git repository.

**Comment by Ericson2314** [comment]
Hmm, `termite` is now marked broken, so I am very interested in seeing this finished :).

**Comment by OJFord** [comment]
I'd also like to use alacritty, but what's the status of this? #21800 has been merged, but I don't think this works regardless?

If I build manually, having added the dependencies to my configuration.nix, I have the issue described in jwilm/alacritty#275 - but the solution there of adding a directory to `$LD_LIBRARY_PATH` is no good, since the directory doesn't exist on my system...

**Comment by grahamc** [comment]
After this applying this patch:

```diff
diff --git a/pkgs/applications/misc/alacritty/default.nix b/pkgs/applications/misc/alacritty/default.nix
index ec312ed..952bcda 100644
--- a/pkgs/applications/misc/alacritty/default.nix
+++ b/pkgs/applications/misc/alacritty/default.nix
@@ -7,13 +7,13 @@ buildRustPackage rec {
   name = "alacritty-unstable-${version}";
   version = "2017-01-07";
 
-  depsSha256 = "138bfnrg3j49vl6j2daf1jyyng5r6d088qhsqikhyyqzjx1ih0v2";
+  depsSha256 = "1yv7nhppixgf62yy3is8wlbcpsgid3qjwp2aaj54c456k4gzhxpd";
 
   src = fetchFromGitHub {
     owner = "jwilm";
     repo = "alacritty";
-    rev = "f4b10a1dbb991861a84c5085a0ee4384d2377448";
-    sha256 = "044l5s1w2g5krngr1v4lfxv7smh6wzy9zrabnihnpf1bsg6q73nb";
+    rev = "59295e44312b3936132965636e5fac92118e5c27";
+    sha256 = "0jk0nyc8v4ky1riyfbv1z52a35zhiaspggmzqxpggbq7vnsbirx1";
   };
 
   nativeBuildInputs = [
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index be9b1f0..78b3564 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -380,7 +380,6 @@ with pkgs;
 
   alacritty = callPackage ../applications/misc/alacritty {
     inherit (xorg) libXcursor libXxf86vm libXi;
-    rustPlatform = (makeRustPlatform rustNightly);
   };
 
   amazon-glacier-cmd-interface = callPackage ../tools/backup/amazon-glacier-cmd-interface { };
diff --git a/pkgs/top-level/rust-packages.nix b/pkgs/top-level/rust-packages.nix
index e35153a..db067a1 100644
--- a/pkgs/top-level/rust-packages.nix
+++ b/pkgs/top-level/rust-packages.nix
@@ -7,9 +7,9 @@
 { runCommand, fetchFromGitHub, git }:
 
 let
-  version = "2017-01-27";
-  rev = "6a73a15e27364a0c191d61d52406bebb7639b657";
-  sha256 = "1dklswbf3acfqid4vv7g2xpqc4gswjgh4ih9xjx3a0m3a69cw9lb";
+  version = "2017-02-07";
+  rev = "e56dbd2ee8dbcc29b8c72586bff64a99268fecf3";
+  sha256 = "0n69ia1r67p9dgrhagrlii9n6wqqasbb1ggma90mawfmvfr7ay7c";
 
   src = fetchFromGitHub {
       inherit rev;
```

I'm seeing:


```
Using cargo deps from /nix/store/0hm0mfx0m7c8477yn1gp1llh8rk9mmi5-alacritty-unstable-2017-01-07-fetch
Using indexHash '-ba82b75dd6681d6f'
Using rust registry from /nix/store/11lvmj2525ri3g3s2xsfggxaq8sb1wsk-rustRegistry-2017-02-07-e56dbd2
warning: custom registry support via the `registry.index` configuration is being removed, this functionality will not work in the future
    Updating git repository `https://github.com/jwilm/glutin`
warning: spurious network error (2 tries remaining): [12/-1] curl error: Couldn't resolve host 'github.com'

warning: spurious network error (1 tries remaining): [12/-1] curl error: Couldn't resolve host 'github.com'

error: failed to load source for a dependency on `glutin`

Caused by:
  Unable to update https://github.com/jwilm/glutin?rev=af7fe340bd4a2af53ea521defcb4f377cdc588cf#af7fe340

Caused by:
  failed to fetch into /tmp/nix-build-alacritty-unstable-2017-01-07.drv-0/deps/git/db/glutin-2b5ecc80feb2e4c5

To learn more, run the command again with --verbose.
builder for ‘/nix/store/hfx0vyils7fi6hp327alv7v8j5pr4p4q-alacritty-unstable-2017-01-07.drv’ failed with exit code 101
error: build of ‘/nix/store/hfx0vyils7fi6hp327alv7v8j5pr4p4q-alacritty-unstable-2017-01-07.drv’ failed
```

**Comment by jgeerds** [comment]
any update?

**Comment by gilligan** [comment]
I didn't check open PRs before packaging it on my own: https://gist.github.com/gilligan/63d4d4832fae996f3225cb59345c689e

It is working just fine for me on NixOS.

**Comment by miltador** [comment]
@gilligan, do you have any progress?

**Comment by trizinix** [comment]
The following works for me: [default.nix](https://gist.github.com/trizinix/10c0e165d71529e916c2103be6a17bc7)
I used stable rust instead of nightly:
```
alacritty = callPackage ../applications/misc/alacritty {
  inherit (xorg) libXcursor libXxf86vm libXi;
};
```

**Comment by grahamc** [comment]
@trizinix Building that with the sandbox enabled, I got:

```
warning: custom registry support via the `registry.index` configuration is being removed, this functionality will not work in the future
    Updating git repository `https://github.com/jwilm/glutin`
warning: spurious network error (2 tries remaining): [12/-1] curl error: Couldn't resolve host 'github.com'

warning: spurious network error (1 tries remaining): [12/-1] curl error: Couldn't resolve host 'github.com'

error: failed to load source for a dependency on `glutin`

Caused by:
  Unable to update https://github.com/jwilm/glutin?rev=cc64178d39a1fa06b2c5403117e5e0ef24deeac4#cc64178d

Caused by:
  failed to fetch into /tmp/nix-build-alacritty-0.1.0.drv-0/deps/git/db/glutin-2b5ecc80feb2e4c5

```

**Comment by trizinix** [comment]
@grahamc This issue is blocked by jwilm/alacritty#234. Alacritty depends on a fork of glutin, as specified in Cargo.toml. Cargo tries to fetch the glutin dependency from github during build time, which fails in the sandbox. As a workaround we could download glutin before the build-phase and then modify Cargo.toml accordingly.

**Comment by therealpxc** [comment]
@trizinix I just tried doing that, but I couldn't get it to work. Does buildRustPackage not handle the `patches` and `postPatchPhase` attributes in a normal way?

**Comment by Mic92** [comment]
@therealpxc cargo runs before the patch phase. You can patch it like this: https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/misc/exa/default.nix#L47

**Comment by therealpxc** [comment]
Thanks, @mic92 . Now I've implemented a temporary workaround on my machine parallel to the one they use in the exa package. But there's another problem: jwilm's custom glutin fork *also* uses git dependencies, so now I get the same failure because glutin depends on jwilm/rust-fontconfig. That version of fontconfig further depends on jwilm/libfontconfig, also specified as a git dependency in its Cargo.toml file.

What can we do to patch the additional sources? Can we still define them using fetchFromGitHub or do we need real derivations now, since we have to patch them?

**Comment by Mic92** [comment]
@therealpxc this was also the showstopper for me. I hope that https://github.com/NixOS/nixpkgs/pull/24991 will handle git support at some point.

**Pull request closed without merging**